tools/autograd/templates/ViewFuncs.h

```diff
@@ -0,0 +1,149 @@
+/// where we want to use symbolic values or fake tensors instead.
+struct TORCH_API ViewFunc {
+  virtual ~ViewFunc() {}
+  /// Returns any SymInts in the saved state.
+  virtual std::vector<c10::SymInt> get_symints() const = 0;
+  /// Returns any Tensors in the saved state.
+  virtual std::vector<at::Tensor> get_tensors() const = 0;
+  /// Reapplies the view on the given base using the saved state.
+  /// If specified, the provided values for SymInts / Tensors
+  /// are hot-swapped in for the view call.
+  virtual at::Tensor operator()(
+      const at::Tensor&,
+      const std::optional<std::vector<c10::SymInt>>& = c10::nullopt,
+      const std::optional<std::vector<at::Tensor>>& = c10::nullopt) const = 0;
+  // Allow limited setter access to maintain the const-ness invariant.
+  friend class ViewFuncSavedStateGuard;
+
+protected:
+  /// Sets the values of any SymInts in the saved state. The input vector size must
+  /// match the number of SymInts in the saved state (i.e. the size of the list
+  /// returned by get_symints()).
+  virtual void set_symints(const std::vector<c10::SymInt>&) = 0;
+  /// Sets the values of any Tensors in the saved state. The input vector size must
+  /// match the number of Tensors in the saved state (i.e. the size of the list
+  /// returned by get_tensors()).
+  virtual void set_tensors(const std::vector<at::Tensor>&) = 0;
+};
+
+// RAII guard for modifying and restoring saved SymInt / tensor state in a
+// ViewFunc.
+struct ViewFuncSavedStateGuard {
+  ViewFuncSavedStateGuard(
+      ViewFunc& view_func,
+      const std::optional<std::vector<c10::SymInt>>& new_symints,
+      const std::optional<std::vector<at::Tensor>>& new_tensors)
+      : view_func(view_func) {
+    // save old state and set new state
+    old_symints = view_func.get_symints();
+    old_tensors = view_func.get_tensors();
+    if (new_symints.has_value()) {
+      view_func.set_symints(*new_symints);
+    }
+    if (new_tensors.has_value()) {
+      view_func.set_tensors(*new_tensors);
+    }
+  }
+
+  ~ViewFuncSavedStateGuard() {
+    // restore previous state
+    view_func.set_symints(old_symints);
+    view_func.set_tensors(old_tensors);
+  }
+
+  ViewFunc& view_func;
+  std::vector<c10::SymInt> old_symints;
+  std::vector<at::Tensor> old_tensors;
+};
+
+struct ChainedViewFunc : public ViewFunc {
+  ChainedViewFunc(
+      const std::shared_ptr<ViewFunc>&,
+      const std::shared_ptr<ViewFunc>&);
+  virtual ~ChainedViewFunc() override {};
+  virtual std::vector<c10::SymInt> get_symints() const override;
+  virtual std::vector<at::Tensor> get_tensors() const override;
+  virtual at::Tensor operator()(
+      const at::Tensor&,
+      const std::optional<std::vector<c10::SymInt>>& = c10::nullopt,
+      const std::optional<std::vector<at::Tensor>>& = c10::nullopt) const override;
+
+protected:
+  // Purposefully do nothing for these and let the sub-ViewFuncs handle it
```

Comment:
Do you want this to be virtual (`= 0`)? Is it an error condition not to define these virtual methods?

Reply:
Revisited this; now the only pure virtual APIs are `operator()` and `clone_and_set()`. The generated ViewFuncs make use of `set_symints()` / `set_tensors()` underneath `clone_and_set()`, but `ChainedViewFunc` doesn't need them.